docs/use_cases/node_representation_learning.md

```diff
@@ -0,0 +1,269 @@
+3. We sample neighbors for a depth of 2 for each selected node. In this sampling process, we won't take into consideration all of the neighbors, instead we'll only sample a fixed size of neighbors, defined by the `num_neighbors` parameter. So in this case in the first hop we are sampling 15 neighbors while only 10 in the second layer. This is particularly useful, since limiting the considered neighbors will decouple computational complexity from the actual node degree.
+
+```python
+from torch_geometric.loader import LinkNeighborLoader
+loader = LinkNeighborLoader(
+    ds,
+    batch_size=1024,
+    shuffle=True,
+    neg_sampling_ratio=1.0,
+    num_neighbors=[15,10],
+    transform=T.NormalizeFeatures(),
+    num_workers=4
+)
+```
+
+Here we can see how a batch, returned by the loader actually looks like:
+
+```python
+print(next(iter(loader)))
+>>> Data(x=[2646, 1433], edge_index=[2, 8642], edge_label_index=[2, 2048], edge_label=[2048], ...)
+```
+In the `Data` object `x` contains the BoW node features, the `edge_label_index` tensor contains the head and tail node indices for the positive and negative samples, `edge_label` is the binary target for these pairs (1 for positive 0 for negative samples). The `edge_index` tensor holds the adjacency list for the current batch of nodes. 
+
+Now we can train our model as follows:
+```python
+def train():
+    sage.train()
+    total_loss = 0
+    for batch in loader:
+        batch = batch.to(device)
+        optimizer.zero_grad()
+        # create node representations
+        h = sage(batch.x, batch.edge_index)
+        # take head and tail representations
+        h_src = h[batch.edge_label_index[0]]
+        h_dst = h[batch.edge_label_index[1]]
+        # compute pairwise edge scores
+        pred = (h_src * h_dst).sum(dim=-1)
+        # apply cross entropy
+        loss = F.binary_cross_entropy_with_logits(pred, batch.edge_label)
+        loss.backward()
+        optimizer.step()
+        total_loss += float(loss) * pred.size(0)
+    return total_loss / ds.num_nodes
+
+for epoch in range(200):
+    loss = train()
+    print(f'Epoch: {epoch:03d}, Loss: {loss:.4f}')
+```
+
+Now that we have the trained model, we can embed nodes and evaluate the embeddings on the classification task:
+
+```python
+embeddings = sage(normalize(ds.x), ds.edge_index).detach().cpu()
+evaluate(embeddings, ds.y)
+>>> Accuracy 0.834
+>>> F1 macro 0.818
+```
+
+The results are slightly worse (3%) than the results we got by combining Node2Vec with BoW features however, remember that with this model we can embed completely new nodes too. If our scenario requires inductiveness, GraphSAGE might be a better solution however, if we had a transductive setting, Node2Vec would give us a better solution.
+
+## Conclusion
```

Comment:
Don't you think it would be worth embedding the text of the papers with some sentence transformer model also? And repeat the scenarios where it is concatenated to node2vec?
GraphSage works on the vectors, or does it embed the text itself? Because it could be worth adding it to that scenario as well. [This](https://huggingface.co/sentence-transformers/all-mpnet-base-v2) is a reasonably sized, relatively well performing model.

Reply:
1. Sure, I will try to do that. Unfortunately the `torch_geometric` dataset does not contain the text of the articles. However, I found the original data (from which the `torch_geometric` dataset should be derived) that contains paper extracts. I will try to match the paper IDs and embed the abstracts with the LLM.
2. GraphSAGE uses the BoW features as input. Also we can try to train the sage model with the LLM features.